### DISCUSSION TITLE
A lot of time is spent in wait4() when running a BATS test-suite

### DISCUSSION THREAD

**Discussion by brgl** [bats-core/bats-core#731] [Q&A]
I'm the author of [libgpiod](https://git.kernel.org/pub/scm/libs/libgpiod/libgpiod.git/) which uses BATS for its command-line tools.

I noticed that these tests - even though they use the same kernel infrastructure in the form of the `gpio-sim` module - take much longer to complete than the unit tests for the core C library. We're talking several orders of magnitude: library tests: 0,28s on my laptop, bash tests: ~26 seconds(!). Some users have even reported that it takes over 2 minutes for the tests to complete. They cover pretty much the same functionality and while it's clear that bash tests will be much slower, I don't think this is entirely normal.

A quick run of `strace -c -f` reveals the following:

```
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ------------------
 91,88   19,167164         944     20285      8653 wait4
  1,72    0,359402           0    814675           rt_sigprocmask
  1,30    0,271654           1    174942    145707 readlink
  1,22    0,254268           0    261146         8 read
  1,02    0,211915          18     11283           clone
  0,50    0,104316          16      6398       665 execve
  0,47    0,097529           0    128670     13692 newfstatat
  0,36    0,074594           0     95369     49559 ioctl
  0,25    0,052272           0     77944           mmap
  0,24    0,050331           2     22081           write
[...]
```

The majority of time is spent in `wait4()` and I'm seeing a lot of `wait4(-1, 0x<snip>, WNOHANG, NULL) = -1 ECHILD (No child processes)` calls in the complete `strace` output.

Any idea what could be causing this?

**Comment by martin-schulze-vireso** [comment]
I think we need more details. In essence this is a bug report so please provide the usual details:

- OS version
- bash version
- Bats version
- Test Code
- Bats invocation command line

**Reply by brgl** [reply]
I wasn't sure if this is a bug or a feature so I opened a discussion, but sure, I'll post a detailed bug report in the issues section.

**Reply by martin-schulze-vireso** [reply]
You are welcome to post this as a bug issue but I Just wanted to request the usual information that is necessary to reproduce this issue.
